File: sample/cpp/06_julia_parameter_sweep.cpp
#include "cpp_module/py_runtime.h"





list<uint8> julia_palette() {
    float64 t;
    int64 r;
    int64 g;
    int64 b;
    
    list<uint8> palette = list<uint8>(256 * 3);
    palette[0] = 0;
    palette[1] = 0;
    palette[2] = 0;
    for (int64 i = 1; i < 256; ++i) {
        t = (static_cast<float64>(i - 1)) / 254.0;
        r = int64(255.0 * 9.0 * (1.0 - t) * t * t * t);
        g = int64(255.0 * 15.0 * (1.0 - t) * (1.0 - t) * t * t);
        b = int64(255.0 * 8.5 * (1.0 - t) * (1.0 - t) * (1.0 - t) * t);
        palette[i * 3 + 0] = r;
        palette[i * 3 + 1] = g;
        palette[i * 3 + 2] = b;
    }
    return list<uint8>(palette);
}

list<uint8> render_frame(int64 width, int64 height, float64 cr, float64 ci, int64 max_iter, int64 phase) {
    float64 zy0;
    float64 zx;
    float64 zy;
    int64 i;
    float64 zx2;
    float64 zy2;
    int64 color_index;
    
    list<uint8> frame = list<uint8>(width * height);
    int64 idx = 0;
    for (int64 y = 0; y < height; ++y) {
        zy0 = -1.2 + 2.4 * static_cast<float64>(y) / (static_cast<float64>(height - 1));
        for (int64 x = 0; x < width; ++x) {
            zx = -1.8 + 3.6 * static_cast<float64>(x) / (static_cast<float64>(width - 1));
            zy = zy0;
            i = 0;
            while (i < max_iter) {
                zx2 = zx * zx;
                zy2 = zy * zy;
                if (zx2 + zy2 > 4.0)
                    break;
                zy = 2.0 * zx * zy + ci;
                zx = zx2 - zy2 + cr;
                i++;
            }
            if (i >= max_iter) {
                frame[idx] = 0;
            } else {
                color_index = 1 + (py_floordiv(i * 224, max_iter) + phase) % 255;
                frame[idx] = color_index;
            }
            idx++;
        }
    }
    return list<uint8>(frame);
}

void run_06_julia_parameter_sweep() {
    float64 t;
    float64 angle;
    float64 cr;
    float64 ci;
    int64 phase;
    
    int64 width = 320;
    int64 height = 240;
    int64 frames_n = 72;
    int64 max_iter = 180;
    str out_path = "sample/out/06_julia_parameter_sweep.gif";
    
    float64 start = perf_counter();
    list<list<uint8>> frames = list<list<uint8>>{};
    
    float64 center_cr = -0.745;
    float64 center_ci = 0.186;
    float64 radius_cr = 0.12;
    float64 radius_ci = 0.1;
    
    
    int64 start_offset = 20;
    int64 phase_offset = 180;
    for (int64 i = 0; i < frames_n; ++i) {
        t = static_cast<float64>((i + start_offset) % frames_n) / static_cast<float64>(frames_n);
        angle = 2.0 * py_math::pi * t;
        cr = center_cr + radius_cr * py_math::cos(angle);
        ci = center_ci + radius_ci * py_math::sin(angle);
        phase = (phase_offset + i * 5) % 255;
        frames.push_back(render_frame(width, height, cr, ci, max_iter, phase));
    }
    
    save_gif(out_path, width, height, frames, julia_palette(), 8, 0);
    float64 elapsed = perf_counter() - start;
    py_print("output:", out_path);
    py_print("frames:", frames_n);
    py_print("elapsed_sec:", elapsed);
}

int main() {
    run_06_julia_parameter_sweep();
    return 0;
}

File: sample/scala/06_julia_parameter_sweep.scala
import scala.collection.mutable
import scala.util.boundary, boundary.break
import scala.math.*
import java.nio.file.{Files, Paths}


// 06: Sample that sweeps Julia-set parameters and outputs a GIF.

def julia_palette(): mutable.ArrayBuffer[Long] = {
    var palette: mutable.ArrayBuffer[Long] = __pytra_bytearray(256L * 3L)
    __pytra_set_index(palette, 0L, 0L)
    __pytra_set_index(palette, 1L, 0L)
    __pytra_set_index(palette, 2L, 0L)
    var i: Long = 1L
    while (i < 256L) {
        var t: Double = ((__pytra_float(i - 1L)) / 254.0)
        var r: Long = __pytra_int(255.0 * ((((9.0 * (1.0 - t)) * t) * t) * t))
        var g: Long = __pytra_int(255.0 * ((((15.0 * (1.0 - t)) * (1.0 - t)) * t) * t))
        var b: Long = __pytra_int(255.0 * ((((8.5 * (1.0 - t)) * (1.0 - t)) * (1.0 - t)) * t))
        __pytra_set_index(palette, (i * 3L + 0L), r)
        __pytra_set_index(palette, (i * 3L + 1L), g)
        __pytra_set_index(palette, (i * 3L + 2L), b)
        i += 1L
    }
    return __pytra_bytes(palette)
}

def render_frame(width: Long, height: Long, cr: Double, ci: Double, max_iter: Long, phase: Long): mutable.ArrayBuffer[Long] = {
    var frame: mutable.ArrayBuffer[Long] = __pytra_bytearray(width * height)
    var __hoisted_cast_1: Double = __pytra_float(height - 1L)
    var __hoisted_cast_2: Double = __pytra_float(width - 1L)
    var y: Long = 0L
    while (y < height) {
        var row_base: Long = y * width
        var zy0: Double = ((-1.2) + (2.4 * (__pytra_float(y) / __hoisted_cast_1)))
        var x: Long = 0L
        while (x < width) {
            var zx: Double = ((-1.8) + (3.6 * (__pytra_float(x) / __hoisted_cast_2)))
            var zy: Double = zy0
            var i: Long = 0L
            boundary:
                given __breakLabel_2: boundary.Label[Unit] = summon[boundary.Label[Unit]]
                while (i < max_iter) {
                    boundary:
                        given __continueLabel_3: boundary.Label[Unit] = summon[boundary.Label[Unit]]
                        var zx2: Double = zx * zx
                        var zy2: Double = zy * zy
                        if (zx2 + zy2 > 4.0) {
                            break(())(using __breakLabel_2)
                        }
                        zy = ((2.0 * zx * zy) + ci)
                        zx = (zx2 - zy2 + cr)
                        i += 1L
                }
            if (i >= max_iter) {
                __pytra_set_index(frame, row_base + x, 0L)
            } else {
                var color_index: Long = (1L + (((__pytra_int(i * 224L / max_iter) + phase)) % 255L))
                __pytra_set_index(frame, row_base + x, color_index)
            }
            x += 1L
        }
        y += 1L
    }
    return __pytra_bytes(frame)
}

def run_06_julia_parameter_sweep(): Unit = {
    var width: Long = 320L
    var height: Long = 240L
    var frames_n: Long = 72L
    var max_iter: Long = 180L
    var out_path: String = "sample/out/06_julia_parameter_sweep.gif"
    var start: Double = __pytra_perf_counter()
    var frames: mutable.ArrayBuffer[Any] = __pytra_as_list(mutable.ArrayBuffer[Any]())
    var center_cr: Double = __pytra_float(-0.745)
    var center_ci: Double = 0.186
    var radius_cr: Double = 0.12
    var radius_ci: Double = 0.1
    var start_offset: Long = 20L
    var phase_offset: Long = 180L
    var __hoisted_cast_3: Double = __pytra_float(frames_n)
    var i: Long = 0L
    while (i < frames_n) {
        var t: Double = (__pytra_float((i + start_offset) % frames_n) / __hoisted_cast_3)
        var angle: Double = __pytra_float(2.0 * Math.PI * t)
        var cr: Double = __pytra_float(center_cr + radius_cr * scala.math.cos(__pytra_float(angle)))
        var ci: Double = __pytra_float(center_ci + radius_ci * scala.math.sin(__pytra_float(angle)))
        var phase: Long = (((phase_offset + i * 5L)) % 255L)
        frames.append(render_frame(width, height, cr, ci, max_iter, phase))
        i += 1L
    }
    __pytra_save_gif(out_path, width, height, frames, julia_palette())
    var elapsed: Double = __pytra_perf_counter() - start
    __pytra_print("output:", out_path)
    __pytra_print("frames:", frames_n)
    __pytra_print("elapsed_sec:", elapsed)
}

def main(args: Array[String]): Unit = {
    run_06_julia_parameter_sweep()
}